### PR TITLE
fix: do not dllexport symbols in core static libraries

### PR DESCRIPTION
**Pull request by Kha**
On Windows, we now compile all core `.o`s twice, once with and without `dllexport`, for use in the shipped dynamic and static libraries, respectively. On other platforms, we export always as before to avoid the duplicate work.

### CODE REVIEW

File: src/lean.mk.in [hunk 1]
```diff
@@ -93,22 +94,33 @@ $(BC_OUT)/%.bc: $(OLEAN_OUT)/%.olean
...
 else
-$(TEMP_OUT)/%.o: $(C_OUT)/%.c
+$(TEMP_OUT)/%.o.export: $(C_OUT)/%.c
 endif
 ifdef CMAKE_LIKE_OUTPUT
 	@echo "[    ] Building $<"
 endif
 	@mkdir -p "$(@D)"
-ifdef PROFILE
+	$(LEANC) -c -o $@ $< $(LEANC_OPTS) -DLEAN_EXPORTING
```
[the remark this answers is not in the document]
Comment (Kha) [reply src/lean.mk.in:105]: Oh yes, I meant to flag this to @bollu @hargoniX but forgot. I don't see this being addressed before we move code generation out of the main `lean` invocation

**Comment by Kha** [comment]
@tydeu I believe the only remaining issue is that `supportInterpreter` test binaries now fail on Windows as we've now basically implemented our own `-rdynamic` distinction! So Lake would need to include `LEANC_SHARED_LINKER_FLAGS` when building a `supportInterpreter` executable on Windows.

**Comment by tydeu** [comment]
@Kha The Lake linking fixed on of the tests, but the other is now failing with `l_System_FilePath_join was replaced` ([log](https://github.com/leanprover/lean4/actions/runs/8192309889/job/22403532585?pr=3601#step:17:3681)).

**Comment by Kha** [comment]
Interesting. I can only assume that means that we try to both statically and dynamically link against this function, but I don't know why that should happen if `-lInit_shared` does land in front of `-lInit`. Would be great if you could locally test what `clang` invocation makes this work.

File: src/runtime/CMakeLists.txt [hunk 2]
```diff
@@ -16,6 +16,10 @@ set_target_properties(leanrt PROPERTIES
 if (NOT MSVC)
 target_compile_options(leanrt PRIVATE -ftls-model=local-exec)
 endif()
+# We do not export Lean symbols when statically linking on Windows.
+if(${CMAKE_SYSTEM_NAME} MATCHES "Windows")
+target_compile_options(leanrt PRIVATE -ULEAN_EXPORTING)
+endif()
```
Comment (tydeu) [inline src/runtime/CMakeLists.txt:22]: @Kha I noticed that executables still exported symbols from the leanrt and leancpp. Fixing this for `leanrt` was relatively trivial (as seen above); however, after some trial and error, it is not clear to me how to fix this for `leancpp` without significant code reorganization (as `leancpp` is composed of nested object libraries and it is not possible to easily build a copy of them with different `-D` switches via CMake). What do you think we should do here? Having no exported symbols seems ideal, as it prevents the possibility of a clash when linking/loading shared libraries like `libleanshared` with the executable.

**Comment by tydeu** [comment]
I believe I have fixed everything I can on the Lake side and everything now can function, so I will hand this back to you @Kha.

File: src/runtime/CMakeLists.txt (hunk 2 above)
Comment (Kha) [reply src/runtime/CMakeLists.txt:22]: > as it prevents the possibility of a clash when linking/loading shared libraries like `libleanshared` with the executable.

I don't understand this part, we should link against leancpp either statically or dynamically, not both

Comment (tydeu) [reply src/runtime/CMakeLists.txt:22]: (Resolved offline.)

**Comment by tydeu** [comment]
@Kha Added the release note. Feel free to merge when convenient.

**Pull request merged**